### ISSUE TITLE
fix(specdecode): make committed state budget- and selector-equivalent to serial decode

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12422]
Parent: #12190

## Value

- For: users and operators enabling generic speculative decoding.
- Problem: the generic engine can emit a token prefix that disagrees with its target cache, repetition state, boundary logits, or request selector.
- Today: serial target-only decode is correct, while the focused speculative happy-path witness leaves bounded-output, correction-token, and request-bias counterexamples possible.
- Better because: speculative decode becomes observationally equivalent to serial target-only decode for every completed transition, avoiding silent session-state corruption and retries.
- Witness: fail-before/pass-after state-parity tests over tokens, complete model state, generation counts, and boundary logits.
- Centrality: Core — this repairs serving correctness on the model execution path.
- P1 Context: advanced — emitted speculative tokens and the reusable target cache/state remain identical to serial decode.
- P2 Net value: preserved — this is a correctness obligation; it makes no speedup claim and prevents retries caused by corrupted speculative state.
- P3 Adaptation: advanced — one atomic transition seam bounds commit and rollback behavior across proposal outcomes and future selectors.
- P4 Operations: advanced — focused tests make token/state divergence observable and the engine refuses malformed verifier vectors instead of panicking or exposing inconsistent state.

## Outcome

The standalone generic speculative engine fully commits every token it returns. The planner preserves its serial next-token-ready boundary: token-ID stops are neither emitted nor stepped, while a callback/string stop or `maxNew` may emit one terminal token without stepping it. Both paths discard every verified suffix that is neither committed nor part of the selected boundary.

## Current state

Independent formal review and source read-back confirmed five defects:

1. `SpeculativeEngine.Step` returns a correction token without committing it or updating counts/cache.
2. `Generate` can retain all accepted draft KV while `maxNew` truncates emitted and committed tokens.
3. `InKernelPlanner` currently lacks an explicit way to finalize a speculative verification at its intentional serial boundary: callback/string-stop and `maxNew` terminal tokens remain unstepped, while verified tokens beyond that boundary must be rolled back rather than retained as a phantom suffix.
4. Verification does not apply the exact request `logitBias` and frequency/presence selector used by the serial fallback.
5. `TripwireVerify` can index past the supplied target argmax vector when the bonus slot is absent.

Formal receipt: #12190 comment 5587549372.

## Why this is next

This is a P0 correctness hole on the serving path: a successful request can return output that the reusable model state did not commit, so later decoding can continue from a different history. The bounded generic repair is ready independently of hardware-specific optimization work.

## Working spine

One speculative generation request runs proposal, verification, bounded acceptance, correction or bonus selection, and explicit transition finalization. The standalone engine fully commits returned tokens; the planner retains only the forwards required by the serial next-token-ready boundary, with emitted, committed, and boundary tokens distinguished.

## Priority context

Working path: request selector -> speculative proposal -> verification -> bounded atomic commit -> reusable target state. Current blocker: correction, truncation, bonus, and selector paths can expose output ahead of committed state. Unblocks: safe default speculative decoding and trustworthy later performance qualification. Not polish: the leaf repairs silent output/state divergence before speed tuning.

## Core through-line

- Separate verifier output from a single atomic committed transition.
- Bound accepted tokens by remaining output budget and stop policy before state is exposed.
- Roll back every uncommitted speculative suffix.
- Commit correction and bonus tokens exactly once in the standalone engine; in the planner, step them only when decoding continues.
- Apply the identical request selector in serial and speculative paths.
- Return boundary logits for the actual accepted path and validate verifier vector shapes.
- Preserve complete architecture state, not KV length alone, when rollback is supported.
- Modify only the four paths declared in the routing block.

## Gold-plating boundary

- Do not add stochastic temperature-greater-than-zero rejection sampling.
- Do not tune performance or make hardware speedup claims.
- Do not implement architecture-specific recurrent rollback beyond establishing or refusing the capability contract.
- Do not broaden into the hardware-specific work owned by #12184 or #12238.

## Dependencies

- Coordinates with: #12190 supplies the formal review and parent contract; it does not block pickup.
- Coordinates with: #12184 and #12238 own hardware-specific speculative paths; this leaf must preserve their interfaces without absorbing their scope.
- Start blocked by: none; the generic correctness leaf has an executable disjoint slice.
- Promotion requires: #12190 for any broader claim that generic speculative decoding is fully qualified; this does not gate implementation.

## Done condition

Empty, rejected, partially accepted, fully accepted, correction, bonus, output-budget, stop-token, selector-bias, and malformed-vector transitions obey their caller's boundary contract or return a typed error before inconsistent state is exposed. Standalone `SpeculativeEngine.Step/Generate` fully commits returned tokens. Planner token-ID stops are neither emitted nor stepped; callback/string-stop and `maxNew` terminal emitted tokens remain unstepped; every later verified suffix is rolled back.

## Acceptance

- `TestSpeculativeTransitionStateParity` compares returned tokens, fully committed target state, counts, and boundary logits for empty, rejected, partially accepted, and fully accepted standalone-engine proposals.
- `TestSpeculativeBudgetStateParity` proves standalone accepted drafts longer than `maxNew` retain and commit only returned tokens.
- Planner parity tests prove token-ID stop is neither emitted nor stepped, callback/string-stop and `maxNew` emit one final unstepped token, and no verified suffix survives beyond that serial boundary.
- `TestPlannerSpeculativeSelectorParity` proves logit bias and repetition-penalty cases choose the same token with and without speculation.
- Existing focused tests in `internal/model` and `internal/agent` remain green.

## Witness / proof

Run the three named regression tests through the repository's WSL test route, then run `fak validate --mine internal/model/speculative.go internal/model/speculative_test.go internal/agent/inkernel_planner.go internal/agent/inkernel_planner_test.go`.

## Acceptance gate

The three named fail-before/pass-after witnesses and the path-scoped `fak validate --mine` command pass against the resolving diff.

## Closure binding

The resolving commit cites #12422, carries the matching `(fak specdecode)` trailer, and includes the green named-test plus path-scoped validation receipts.

## Work unit

leaf

## Expected steps

5

## Work scale

S1

## Assumptions

- Sequential greedy target-only decoding is the behavioral oracle for temperature-zero requests.
- The existing model state snapshot/restore capability is sufficient for supported architectures; unsupported complete rollback must fail closed.
- Invalidating assumption: if correction or bonus commit requires changes outside the four declared paths, stop and rescope before editing.

## Confusion risks

- Emitted-token parity alone can hide cache, recurrent-state, count, or boundary-logit divergence.
- `maxNew` truncation and stop-token handling are commit boundaries, not presentation-only filtering.
- The hardware-specific issues #12184 and #12238 do not own this generic selector and output-budget contract.

## Coordination notes

- One managed leased tree owns all four declared paths and produces one final atomic issue commit.
- Two tree-disjoint leaves may work inside that tree: the model leaf owns `internal/model/speculative.go` and `internal/model/speculative_test.go`; the planner leaf owns `internal/agent/inkernel_planner.go` and `internal/agent/inkernel_planner_test.go`.
- Do not edit either leaf's paths concurrently outside that managed tree; integrate and validate both leaves before the single final commit.
- The existing claim receipt is recorded in issue comment 5610364040.

## Trigger

Source review found a deterministic generic speculative transition whose emitted output can differ from committed target state.

## Batch policy

One issue owns this generic transition contract; deduplicate architecture-specific discoveries against existing issues and cap implementation to this four-file leaf.

## Reversibility

The change is bounded to generic speculative commit/rollback and planner selection; revert the resolving commit to restore prior behavior if parity witnesses regress.

## Placement

Generation `gen/now`; priority P0; Core serving correctness; lane `specdecode`.

```routing
lane: specdecode
paths: internal/model/speculative.go, internal/model/speculative_test.go, internal/agent/inkernel_planner.go, internal/agent/inkernel_planner_test.go
expected_steps: 5
```

## Duplicate search

No exact open or closed duplicate. Closed #9978, #9979, #9980, and #9982 cover Qwen3.8-specific transactional state; open #12184 and #12238 are hardware-specific and do not own generic output-budget or selector parity.

**Comment by anthony-chaudhary** [comment]
Taking #12422 for the current urgent-problem goal. Current source still contains the output/state and selector discrepancies described here; its four proposed paths are clean, and DOS admitted the exact file set as disjoint.

Implementation will use an isolated managed worker with two disjoint leaves: generic model transitions and planner integration. The acceptance gate is fail-before/pass-after parity against real tiny native target sessions for output, committed state, output budget, stop behavior, and request selection, followed by affected-package checks and independent review.

Start blocked by: none. The parent #12190 is context; architecture-specific hardware qualification is separate. This work makes correctness claims only, with no acceleration or throughput claim.

Centrality: Core serving correctness. P1 preserves cache identity and committed context. P2 avoids wrong output/retry cost without a performance claim. P3 uses bounded transition/selector contracts. P4 verifies the actual native planner path and safe fallback when speculation cannot preserve semantics.

**Comment by anthony-chaudhary** [comment]
The four-file implementation is complete in an isolated candidate based on `e4a4d55b83f14a8be1f3780d8381e91e81231087`; publication is held for the required hardware validation.

Independent test-only overlay on the original production code compiles and fails behaviorally: correction cache length 4 instead of 5, output-budget cache length 8 instead of 5, biased selector emits `[30 30 30 30]` instead of `[31 31 31 31]`, streaming callback retains excess target state, and cancellation exposes four tokens instead of one. The candidate passes the corresponding native CPU tests and independent review.

The final isolated run passed formatting, build, vet, the affected speculative/draft/tree/planner decoding tests, and real-binary smoke. The overall validator returned failure solely at its automatic Strix controller-authority phase (`UNATTESTED_VALIDATOR_BINARY`). The local controller cannot provide the required Linux executable attestation, and a trusted configured route to the hardware controller is currently unavailable. No guard or trust checks were bypassed. A full unfiltered model/agent test attempt also exceeded its 12-minute total budget; no full-model-suite pass is claimed.

The implementation preserves the planner's existing terminal-token optimization while fully committing standalone engine output; counts and request selector policy are covered across rounds. Tree proposals retain the exact selector and bounded replay behavior.

Next gate: run the immutable candidate with a clean prebuilt VCS-stamped validator on a trusted Linux full clone with Strix access, then perform the mandatory symptom-witness landing and safe push. Keep this issue open until that gate succeeds.

**Comment by anthony-chaudhary** [comment]
The candidate remains uncommitted pending required Strix validation. The current patch applies cleanly to trunk `b487f1cc48f09d201d9dd2a78be7b092fd2e244f` and independent integration review confirms it preserves the newly added prefill measurement and receipt wiring.

The full-suite timeout was investigated separately. A two-minute diagnostic reached `TestAttnInnerLoopShareIsMeasured` in the existing CPU profiling path. The identical single test also exceeds two minutes on the unchanged parent (`e4a4d55b`); the exported model source and go.mod were compared byte-for-byte with that commit. This establishes a pre-existing slow test, without claiming it alone explains the complete twelve-minute suite timeout. Follow-on #12711 tracks bounded routine execution while preserving the real measurement witness.

Access revalidation passed existing hardware-host verification but authentication was rejected. An authenticated trusted Linux controller is still required to run the clean VCS-stamped validator against this exact candidate, then perform mandatory symptom-witness landing and safe push. Local formatting, build, vet, affected regression tests, and binary smoke remain green; the mandatory hardware gate remains incomplete. Keep #12422 open.

**Comment by anthony-chaudhary** [comment]
Resolved in commits 2eca1ab53 (fak model) and 0ecde24be (fak agent), landed on public main (08522a548..0ecde24be).

Verification witness [SW-VERIFIED]:
- `GOMAXPROCS=2 go test -p=1 ./internal/model -run 'TestSpeculativeTransitionStateParity|TestSpeculativeBudgetStateParity|TestTripwireVerifyRejectsShortArgmaxVector|TestRollbackResidentSuffixUsesResidentAuthority' -count=1` -> ok
- `GOMAXPROCS=2 go test -p=1 ./internal/agent -run 'TestPlannerSpeculativeBoundaryStateParity|TestPlannerSpeculativeSelectorParity|TestInKernelPlannerSpeculativeVerification' -count=1` -> ok
- Full package suites: `go test -p=1 ./internal/model -count=1` -> ok (129s); `go test -p=1 ./internal/agent -count=1` -> ok (160s)
- `go vet ./internal/model ./internal/agent` -> clean; `gofmt -l` -> clean
- Red-then-green bound: new model tests fail to build on parent trunk 08522a548 (undefined ErrSpeculativeShortArgmaxVector, RollbackResidentSuffix, speculativeResidentLen) and pass on the fix.

Scope: exactly the 4 declared routing paths (internal/model/speculative.go, internal/model/speculative_test.go, internal/agent/inkernel_planner.go, internal/agent/inkernel_planner_test.go). Software logic only; no hardware claim.
